patches/sitecustomize.py

```diff
@@ -1,22 +1,56 @@
 # TODO(rosbo): Remove this once we fix the issue with fastai importing older libcudnn if imported prior to tensorflow
 import tensorflow
 import os
+from google.auth import credentials
+from google.cloud import bigquery
+from google.cloud.bigquery._http import Connection
 
-# Monkey patches BigQuery client creation to use proxy.
-kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
-if kaggle_proxy_token:
-    from google.auth import credentials
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+class KaggleKernelCredentials(credentials.Credentials):
+    def refresh(self, request):
+        print("Calling Kaggle.UserSecrets to refresh token.")
+        # Set self.token and self.expiry here.
+        raise NotImplementedError("Private BigQuery integration is not yet implemented.")
 
-    Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
-    Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+CONNECTION_BASE_URL = Connection.API_BASE_URL
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
```

Comment:
For the existing use case, `specified_project` will be set to `None` so it will enter the else block in 41. This means, the existing use case is broken right? Am I missing something?

Reply:
If specified_project is None AND kaggle_proxy_token is None (which should be the case, for the existing use case, since that will only be populated for Kernels where the user has connected an account) then we should still use the proxy.

It's a little confusing but it should be something like this:
 if (kernel has no connected gcp account, ie. no jwt present):
      -  if (user specified a project that's not 'kaggle') : throw an error, since that won't work anyways (today this also already fails)
      - else if (user specified 'kaggle' project or no project): existing behaviour, use data proxy.
if (kernel does have connected gcp account):
       - if (use specifies 'kaggle' project): they are indicating they want to use the proxy explicitly.
        - else : proceed using their connected account instead of proxy (whether or not they specified a proxy).

Reply:
Currently, `KAGGLE_DATA_PROXY_TOKEN` is set to a value to allow access to public BQ as far as I know:
https://www.kaggle.com/rosebv/kaggle-data-proxy-token?scriptVersionId=8721257

This means `kaggle_proxy_token` will be set right?

Reply:
Sorry I meant : If specified_project is None AND *bq_user_jwt* is None (which should be the case, for the existing use case, since that will only be populated for Kernels where the user has connected an account) then we should still use the proxy.

Yeah, kaggle_proxy_token  should always be set.

Reply:
👍